rust/lance/src/dataset/transaction.rs

```diff
@@ -432,16 +454,22 @@ impl Transaction {
                         Some(f.clone())
                     }
                 }));
-                final_fragments.extend(Self::fragments_with_ids(
-                    new_fragments.clone(),
-                    &mut fragment_id,
-                ));
+                let mut new_fragments =
+                    Self::fragments_with_ids(new_fragments.clone(), &mut fragment_id)
+                        .collect::<Vec<_>>();
+                if let Some(next_row_id) = &mut next_row_id {
+                    Self::assign_row_ids(next_row_id, new_fragments.as_mut_slice())?;
+                }
+                final_fragments.extend(new_fragments);
             }
             Operation::Overwrite { ref fragments, .. } => {
-                final_fragments.extend(Self::fragments_with_ids(
-                    fragments.clone(),
-                    &mut fragment_id,
-                ));
+                let mut new_fragments =
+                    Self::fragments_with_ids(fragments.clone(), &mut fragment_id)
+                        .collect::<Vec<_>>();
+                if let Some(next_row_id) = &mut next_row_id {
```

Comment:
Wouldn't we reset row id to 0 on an overwrite?

Reply:
I'm not sure. This will eventually be a stable primary key. I'm not sure it's desireable to re-use ids for different records, even if we are resetting the table with an overwrite. Though I suppose the commit conflict rules would prevent any transactions that might mix things up from happening in the first place. :shrug: